docs/creating-sql-with-python.md

```diff
@@ -0,0 +1,201 @@
+nav_order: 4
+# audience: clinical researcher or engineer familiar with project
+# type: tutorial
+---
+
+# Creating SQL with python
+
+Before jumping into this doc, take a look at 
+[Creating Studies](creating-studies.md).
+If you're just working on `core` tables related to the US Core FHIR profiles, you 
+may not be interested in this, or only need to look at the 
+[Working with TableBuilders](#working-with-tablebuilders)
+and the
+[Generating count tables](#generating-counts-tables)
+sections.
+
+## Why would I even need to think about this?
+
+There are three main reasons why you would need to use python to generate sql:
+- You would like to make use of the 
+[helper class we've built](#generating-counts-tables)
+for ease of creating count tables in a structured manual.
+- You have a dataset you'd like to 
+[load into a table from a static file](#adding-a-static-dataset),
+separate from the ETL tables.
+- The gnarly one: you are working against the raw FHIR resource tables, and are 
+trying to access 
+[nested data](#querying-nested-data) in Athena. 
+  - We infer datatypes in the ETL based on the presence of data once we get past 
+  the top level elements, and so the structure may vary depending on the
+  implementation, either at the EHR level or at the FHIR interface level.
+
+
+We've got examples of all three of these cases in this repo, and we'll reference
+those as examples as we go.
+
+## Utilities
+
+There are two main bits of infrastructure we use for programmatic tables:
+The TableBuilder class, and the collection of template SQL.
+
+### Working with TableBuilders
+
+We have a base
+[TableBuilder class](../cumulus_library/base_table_builder.py)
+that
+all the above use cases leverage. At a high level, here's what it provides:
+
+- A `prepare_queries` function, which is where you put your custom logic. It
+should create an array of queries in `self.queries`. The CLI will pass in a cursor
+object and database/schema name, so if you need to interrogate the dataset to decide
+how to structure your queries, you can.
+- An `execute_queries` function, which will run `prepare_queries` and then apply
+those queries to the database. Generally, you shouldn't need to touch this function -
+just be aware this is how your queries actually get run.
+- A `write_queries` function, which will write your queries from `prepare_function`
+to disk. If you are creating multiple queries in one go, calling `comment_queries`
+before `write_queries` will insert some spacing elements for readability.
+- A `display_text` string, which is what will be shown with a progress bar when your
+queries are being executed.
+
+You can either extend this class directly (like `builder_*.py` files in 
+`cumulus_library/studies/core`) or create a specific class to add reusable functions
+for a repeated use case (like in `cumulus_library/schema/counts.py`).
+
+TableBuilder SQL generally should go through a template SQL generator, so that
+your SQL has been validated.
```

Comment:
nit: For a brief second, I was like "that doesn't seem very helpful - how am I supposed to know how to do that?" before seeing the next header. Maybe "Read below for how to do that." or "In the next section," or some sort of lead-in.

Reply:
added with a caveat about who needs to care.